# Análisis de la Calidad del Aire en el Valle de México

## Introducción

La Organización Mundial de la Salud estima que en 2012 aconteció un aproximado de 3.7 millones de muertes prematuras derivadas de enfermedades atribuidas a la contaminación del aire. Esto se refleja en su discurso del 2014: “La contaminación del aire es el riesgo para la salud ambiental número uno del mundo”.

  

La Ciudad de México, centro político, histórico y financiero de nuestro país cuenta con poco más de 9 millones de habitantes convirtiéndose en la ciudad más poblada de norteamérica, con una densidad de 600,000 habitantes por kilómetro cuadrado. Esto sin contar la totalidad de población que habita la Zona Metropolitana del Valle de México, que se contabiliza por encima de los 21 millones de habitantes.

  

Los problemas que aquejan a la incomparable urbanización no son pocos, desde crimen organizado, alcoholismo, inundaciones periódicas, fríos intensos y el que concierne a esta investigación, la contaminación del aire. De las majestuosas vistas que los primeros españoles hallaron en aquel valle, no queda nada. Incluso el propio cielo se esconde entre una densa nube de vapores tóxicos.

  

Siendo conscientes de este problema, desde el año de 1986, se monitorea de forma periódica la calidad del aire en la Zona Metropolitana del Valle de México por medio del Sistema de Monitoreo Atmosférico (SIMAT) y la RAMA (Red Automática de Monitoreo Atmosférico), realizando mediciones de la concentración de los contaminantes criterio cada hora.

  

Ahora bien, los contaminantes criterio son aquellos que afectan el bienestar y la salud humana, por lo que cuentan con criterios para establecer o revisar límites máximos permisibles por medio de las Normas Oficiales Mexicanas. La concentración de estos contaminantes sirve para indicar la calidad del aire.

  

Dentro del grupo de contaminantes criterio se encuentran el dióxido de azufre (SO2), el dióxido de nitrógeno (NO2), el monóxido de carbono (CO), el ozono (O3) y las partículas en suspensión (PM10 y PM2.5).

Estos contaminantes tienen graves afectaciones a la salud por lo que es necesario no solo monitorear su concentración, sino también obtener con ellos un índice que nos permita clasificar la calidad del aire para así comunicarlo a la población en general. Con este objetivo en mente, en 1982 se diseñó el Índice Metropolitano de la Calidad del Aire (IMECA), cuya metodología transforma a una escala adimensional las concentraciones de los contaminantes criterio; así podemos entender que tan buena o mala es la calidad del aire observando el valor IMECA de dicho contaminante e incluso compararlo con otro, ya que tienen la misma escala.

![](https://lh5.googleusercontent.com/07HClik8fpzXOVvVHCpbBIsc_ku2rtnSCv2p07o0JHnHgUkJp02HyDIQXY0dt8j-ISiy0Rh-wJ-YaKGbYCaPjuk2qQiNh-puh-JleOWb-Fj4z-xp9CrghqKhtReMHtEGzXDHun8a)

                                              Clasificación IMECA
                                              
Durante nuestra investigación nos dimos cuenta que la información sobre la calidad del aire en la Zona Metropolitana del Valle de México no es tan accesible: existen los datos y están abiertos al público, sin embargo, hay muy poca difusión sobre sus resultados, el último informe anual de la calidad del aire en ciudad de méxico se realizó en el 2018. Por lo tanto el objetivo de este proyecto es primero que nada analizar los datos disponibles para entender la problemática de la contaminación del aire en el Valle de México, y hacer que esta información sea más accesible, para luego proponer posibles soluciones.

## Preguntas de Investigación

Planteamos las siguientes preguntas para ayudar a entender y resolver el problema planteado:

-   ¿Cuál es el promedio histórico (considerando los años 2005-2020) de concentración de cada contaminante por zona?

    
-   ¿En qué meses se tiene la máxima concentración de cada contaminante?

    
-   ¿Hay algún mes de algún año en el que se haya obtenido la calificación de calidad “Extremadamente mala”?

    
-   ¿Cuál es la zona que presenta mayor contaminación, en general?

    
-   ¿Cuál es el contaminante con más concentración por año (del 2005- 2020)?

    
-   ¿Cuál fue y cuándo se registró la concentración más baja durante 2020?

    
-   ¿Cuál fue y cuándo se registró la mayor concentración durante el periodo de confinamiento?

    
-   ¿El periodo de confinamiento coincide con una mejor calidad de aire?

    
-   ¿Cuál fue el contaminante con menores cambios en el período 2005-2020?

## Módulo: Procesamiento de Datos

### Lectura y Análisis exploratorio de los datos

In [1]:
#Importando paquetes
import pandas as pd
import numpy as np

In [2]:
#Leyendo un archivo de datos IMECA para averiguar que contiene
df=pd.read_csv("datos_IMECA/csv/imeca2005.csv")
df

Fecha  Hora  Noroeste Ozono  Noroeste dioxido de azufre  \
0     01/01/2005     1              19                          19   
1     01/01/2005     2              19                          19   
2     01/01/2005     3              17                          19   
3     01/01/2005     4              16                          19   
4     01/01/2005     5              16                          19   
5     01/01/2005     6              16                          20   
6     01/01/2005     7              16                          21   
7     01/01/2005     8              14                          22   
8     01/01/2005     9              19                          22   
9     01/01/2005    10              46                          22   
10    01/01/2005    11              74                          23   
11    01/01/2005    12             103                          23   
12    01/01/2005    13             119                          23   
13    01/01/2005    14             122                          23   
14    01/01/2005    15             114                          23   
15    01/01/2005    16             102                          22   
16    01/01/2005    17              75                          22   
17    01/01/2005    18              53                          22   
18    01/01/2005    19              21                          22   
19    01/01/2005    20              25                          22   
20    01/01/2005    21              52                          22   
21    01/01/2005    22              57                          22   
22    01/01/2005    23              49                          22   
23    01/01/2005    24              36                          23   
24    02/01/2005     1              22                          24   
25    02/01/2005     2              34                          23   
26    02/01/2005     3              27                          22   
27    02/01/2005     4              26                          22   
28    02/01/2005     5              26                          21   
29    02/01/2005     6              20                          20   
...          ...   ...             ...                         ...   
8730  30/12/2005    19              23                           5   
8731  30/12/2005    20              12                           5   
8732  30/12/2005    21               9                           5   
8733  30/12/2005    22               9                           5   
8734  30/12/2005    23               9                           5   
8735  30/12/2005    24              10                           6   
8736  31/12/2005     1              10                           6   
8737  31/12/2005     2              11                           6   
8738  31/12/2005     3              14                           7   
8739  31/12/2005     4              18                           7   
8740  31/12/2005     5              15                           8   
8741  31/12/2005     6              12                           8   
8742  31/12/2005     7               9                           8   
8743  31/12/2005     8               9                           8   
8744  31/12/2005     9              12                           8   
8745  31/12/2005    10              21                           9   
8746  31/12/2005    11              42                           9   
8747  31/12/2005    12              68                           9   
8748  31/12/2005    13              85                          10   
8749  31/12/2005    14              94                          10   
8750  31/12/2005    15              95                          10   
8751  31/12/2005    16              92                          10   
8752  31/12/2005    17              55                          10   
8753  31/12/2005    18              35                          10   
8754  31/12/2005    19              22                          10   
8755  31/12/2005    2

In [3]:
#Revisando dimensión
df.shape

(8760, 27)

In [4]:
#Revisando columnas que contiene
df.columns

Index(['Fecha', 'Hora', 'Noroeste Ozono', 'Noroeste dioxido de azufre',
       'Noroeste dioxido de nitrogeno', 'Noroeste monoxido de carbono',
       'Noroeste PM10', 'Noreste Ozono', 'Noreste dioxido de azufre',
       'Noreste dioxido de nitrogeno', 'Noreste monoxido de carbono',
       'Noreste PM10', 'Centro Ozono', 'Centro dioxido de azufre',
       'Centro dioxido de nitrogeno', 'Centro monoxido de carbono',
       'Centro PM10', 'Suroeste Ozono', 'Suroeste dioxido de azufre',
       'Suroeste dioxido de nitrogeno', 'Suroeste monoxido de carbono',
       'Suroeste PM10', 'Sureste Ozono', 'Sureste dioxido de azufre',
       'Sureste dioxido de nitrogeno', 'Sureste monoxido de carbono',
       'Sureste PM10'],
      dtype='object')

Tenemos un archivo de datos por año, vamos a unir todos en un solo DataFrame para trabajar con ellos más facilmente

In [5]:
#primero creamos una lista que contenga todas las rutas de acceso a los archivos csv
lista=[ f'datos_IMECA/csv/imeca{2005+i}.csv' for i in range(0,16)]
lista

['datos_IMECA/csv/imeca2005.csv',
 'datos_IMECA/csv/imeca2006.csv',
 'datos_IMECA/csv/imeca2007.csv',
 'datos_IMECA/csv/imeca2008.csv',
 'datos_IMECA/csv/imeca2009.csv',
 'datos_IMECA/csv/imeca2010.csv',
 'datos_IMECA/csv/imeca2011.csv',
 'datos_IMECA/csv/imeca2012.csv',
 'datos_IMECA/csv/imeca2013.csv',
 'datos_IMECA/csv/imeca2014.csv',
 'datos_IMECA/csv/imeca2015.csv',
 'datos_IMECA/csv/imeca2016.csv',
 'datos_IMECA/csv/imeca2017.csv',
 'datos_IMECA/csv/imeca2018.csv',
 'datos_IMECA/csv/imeca2019.csv',
 'datos_IMECA/csv/imeca2020.csv']

In [6]:
#Realizamos la lectura de todos los archivos csv juntos y reemplazamos -99 por NaN
DF_list = [(pd.read_csv(f'{i}',sep=',')).replace(-99, np.nan) for i in lista]

In [7]:
#comprobamos que se puedan leer de forma correcta los datos
DF_list[0]

Fecha  Hora  Noroeste Ozono  Noroeste dioxido de azufre  \
0     01/01/2005     1              19                          19   
1     01/01/2005     2              19                          19   
2     01/01/2005     3              17                          19   
3     01/01/2005     4              16                          19   
4     01/01/2005     5              16                          19   
5     01/01/2005     6              16                          20   
6     01/01/2005     7              16                          21   
7     01/01/2005     8              14                          22   
8     01/01/2005     9              19                          22   
9     01/01/2005    10              46                          22   
10    01/01/2005    11              74                          23   
11    01/01/2005    12             103                          23   
12    01/01/2005    13             119                          23   
13    01/01/2005    14             122                          23   
14    01/01/2005    15             114                          23   
15    01/01/2005    16             102                          22   
16    01/01/2005    17              75                          22   
17    01/01/2005    18              53                          22   
18    01/01/2005    19              21                          22   
19    01/01/2005    20              25                          22   
20    01/01/2005    21              52                          22   
21    01/01/2005    22              57                          22   
22    01/01/2005    23              49                          22   
23    01/01/2005    24              36                          23   
24    02/01/2005     1              22                          24   
25    02/01/2005     2              34                          23   
26    02/01/2005     3              27                          22   
27    02/01/2005     4              26                          22   
28    02/01/2005     5              26                          21   
29    02/01/2005     6              20                          20   
...          ...   ...             ...                         ...   
8730  30/12/2005    19              23                           5   
8731  30/12/2005    20              12                           5   
8732  30/12/2005    21               9                           5   
8733  30/12/2005    22               9                           5   
8734  30/12/2005    23               9                           5   
8735  30/12/2005    24              10                           6   
8736  31/12/2005     1              10                           6   
8737  31/12/2005     2              11                           6   
8738  31/12/2005     3              14                           7   
8739  31/12/2005     4              18                           7   
8740  31/12/2005     5              15                           8   
8741  31/12/2005     6              12                           8   
8742  31/12/2005     7               9                           8   
8743  31/12/2005     8               9                           8   
8744  31/12/2005     9              12                           8   
8745  31/12/2005    10              21                           9   
8746  31/12/2005    11              42                           9   
8747  31/12/2005    12              68                           9   
8748  31/12/2005    13              85                          10   
8749  31/12/2005    14              94                          10   
8750  31/12/2005    15              95                          10   
8751  31/12/2005    16              92                          10   
8752  31/12/2005    17              55                          10   
8753  31/12/2005    18              35                          10   
8754  31/12/2005    19              22                          10   
8755  31/12/2005    2

In [8]:
#segunda comprobación
DF_list[13]

Fecha  Hora  Noroeste Ozono  Noroeste dioxido de azufre  \
0     01/01/2018     1            24.0                           8   
1     01/01/2018     2            23.0                           8   
2     01/01/2018     3            18.0                           8   
3     01/01/2018     4            12.0                          10   
4     01/01/2018     5            11.0                          12   
5     01/01/2018     6             8.0                          15   
6     01/01/2018     7             9.0                          16   
7     01/01/2018     8            11.0                          18   
8     01/01/2018     9            11.0                          18   
9     01/01/2018    10            19.0                          20   
10    01/01/2018    11            26.0                          22   
11    01/01/2018    12            36.0                          22   
12    01/01/2018    13            41.0                          23   
13    01/01/2018    14            57.0                          23   
14    01/01/2018    15            55.0                          23   
15    01/01/2018    16            50.0                          23   
16    01/01/2018    17            48.0                          23   
17    01/01/2018    18            30.0                          22   
18    01/01/2018    19            17.0                          22   
19    01/01/2018    20            11.0                          22   
20    01/01/2018    21             9.0                          22   
21    01/01/2018    22             9.0                          23   
22    01/01/2018    23             9.0                          24   
23    01/01/2018    24            11.0                          24   
24    02/01/2018     1             7.0                          24   
25    02/01/2018     2             7.0                          24   
26    02/01/2018     3             5.0                          23   
27    02/01/2018     4             9.0                          21   
28    02/01/2018     5             7.0                          18   
29    02/01/2018     6             1.0                          17   
...          ...   ...             ...                         ...   
8730  30/12/2018    19            21.0                           8   
8731  30/12/2018    20            16.0                           8   
8732  30/12/2018    21            18.0                           8   
8733  30/12/2018    22            15.0                           8   
8734  30/12/2018    23            13.0                           8   
8735  30/12/2018    24            14.0                           7   
8736  31/12/2018     1             9.0                           7   
8737  31/12/2018     2             9.0                           5   
8738  31/12/2018     3            11.0                           5   
8739  31/12/2018     4             9.0                           5   
8740  31/12/2018     5             7.0                           5   
8741  31/12/2018     6             2.0                           5   
8742  31/12/2018     7             1.0                           4   
8743  31/12/2018     8             1.0                           4   
8744  31/12/2018     9             4.0                           4   
8745  31/12/2018    10             9.0                           5   
8746  31/12/2018    11            19.0                           5   
8747  31/12/2018    12            27.0                           5   
8748  31/12/2018    13            37.0                           5   
8749  31/12/2018    14            46.0                           5   
8750  31/12/2018    15            48.0                           5   
8751  31/12/2018    16            46.0                           5   
8752  31/12/2018    17            57.0                           5   
8753  31/12/2018    18            50.0                           5   
8754  31/12/2018    19            29.0                           5   
8755  31/12/2018    2

In [9]:
#El Dataframe número 14 y 15 tienen 32 columnas cuando lo correcto serían 27
#Eliminamos las columnas que sobran
cols = [7,13,19,25,31]
DF_list[14]=DF_list[14].drop(DF_list[14].columns[cols],axis=1)
DF_list[15]=DF_list[15].drop(DF_list[15].columns[cols],axis=1)

In [10]:
#Se puede observar que existe un inconveniente: por alguna razón algunos DataFrames 
#tienen columnas de más, que contienen solo elementos vacíos, es necesario eliminarlas
DF_list[12].columns

Index(['Fecha', 'Hora', 'Noroeste Ozono', 'Noroeste dioxido de azufre',
       'Noroeste dioxido de nitrogeno', 'Noroeste monoxido de carbono',
       'Noroeste PM10', 'Noreste Ozono', 'Noreste dioxido de azufre',
       'Noreste dioxido de nitrogeno', 'Noreste monoxido de carbono',
       'Noreste PM10', 'Centro Ozono', 'Centro dioxido de azufre',
       'Centro dioxido de nitrogeno', 'Centro monoxido de carbono',
       'Centro PM10', 'Suroeste Ozono', 'Suroeste dioxido de azufre',
       'Suroeste dioxido de nitrogeno', 'Suroeste monoxido de carbono',
       'Suroeste PM10', 'Sureste Ozono', 'Sureste dioxido de azufre',
       'Sureste dioxido de nitrogeno', 'Sureste monoxido de carbono',
       'Sureste PM10', 'Unnamed: 27', 'Unnamed: 28'],
      dtype='object')

In [11]:
DF_list[12].shape

(8760, 29)

In [12]:
DF_list[12]['Unnamed: 27'] #Se observa que contiene puros elementos vacíos

0      NaN
1      NaN
2      NaN
3      NaN
4      NaN
5      NaN
6      NaN
7      NaN
8      NaN
9      NaN
10     NaN
11     NaN
12     NaN
13     NaN
14     NaN
15     NaN
16     NaN
17     NaN
18     NaN
19     NaN
20     NaN
21     NaN
22     NaN
23     NaN
24     NaN
25     NaN
26     NaN
27     NaN
28     NaN
29     NaN
        ..
8730   NaN
8731   NaN
8732   NaN
8733   NaN
8734   NaN
8735   NaN
8736   NaN
8737   NaN
8738   NaN
8739   NaN
8740   NaN
8741   NaN
8742   NaN
8743   NaN
8744   NaN
8745   NaN
8746   NaN
8747   NaN
8748   NaN
8749   NaN
8750   NaN
8751   NaN
8752   NaN
8753   NaN
8754   NaN
8755   NaN
8756   NaN
8757   NaN
8758   NaN
8759   NaN
Name: Unnamed: 27, Length: 8760, dtype: float64

In [13]:
#ahora eliminamos las columnas que aparecen como sobrantes
DF_list2=[i.dropna(axis="columns", how='all') for i in DF_list]
DF_list2[0]

Fecha  Hora  Noroeste Ozono  Noroeste dioxido de azufre  \
0     01/01/2005     1              19                          19   
1     01/01/2005     2              19                          19   
2     01/01/2005     3              17                          19   
3     01/01/2005     4              16                          19   
4     01/01/2005     5              16                          19   
5     01/01/2005     6              16                          20   
6     01/01/2005     7              16                          21   
7     01/01/2005     8              14                          22   
8     01/01/2005     9              19                          22   
9     01/01/2005    10              46                          22   
10    01/01/2005    11              74                          23   
11    01/01/2005    12             103                          23   
12    01/01/2005    13             119                          23   
13    01/01/2005    14             122                          23   
14    01/01/2005    15             114                          23   
15    01/01/2005    16             102                          22   
16    01/01/2005    17              75                          22   
17    01/01/2005    18              53                          22   
18    01/01/2005    19              21                          22   
19    01/01/2005    20              25                          22   
20    01/01/2005    21              52                          22   
21    01/01/2005    22              57                          22   
22    01/01/2005    23              49                          22   
23    01/01/2005    24              36                          23   
24    02/01/2005     1              22                          24   
25    02/01/2005     2              34                          23   
26    02/01/2005     3              27                          22   
27    02/01/2005     4              26                          22   
28    02/01/2005     5              26                          21   
29    02/01/2005     6              20                          20   
...          ...   ...             ...                         ...   
8730  30/12/2005    19              23                           5   
8731  30/12/2005    20              12                           5   
8732  30/12/2005    21               9                           5   
8733  30/12/2005    22               9                           5   
8734  30/12/2005    23               9                           5   
8735  30/12/2005    24              10                           6   
8736  31/12/2005     1              10                           6   
8737  31/12/2005     2              11                           6   
8738  31/12/2005     3              14                           7   
8739  31/12/2005     4              18                           7   
8740  31/12/2005     5              15                           8   
8741  31/12/2005     6              12                           8   
8742  31/12/2005     7               9                           8   
8743  31/12/2005     8               9                           8   
8744  31/12/2005     9              12                           8   
8745  31/12/2005    10              21                           9   
8746  31/12/2005    11              42                           9   
8747  31/12/2005    12              68                           9   
8748  31/12/2005    13              85                          10   
8749  31/12/2005    14              94                          10   
8750  31/12/2005    15              95                          10   
8751  31/12/2005    16              92                          10   
8752  31/12/2005    17              55                          10   
8753  31/12/2005    18              35                          10   
8754  31/12/2005    19              22                          10   
8755  31/12/2005    2

In [14]:
#observamos que las columnas sobrantes se fueron
DF_list2[12].columns

Index(['Fecha', 'Hora', 'Noroeste Ozono', 'Noroeste dioxido de azufre',
       'Noroeste dioxido de nitrogeno', 'Noroeste monoxido de carbono',
       'Noroeste PM10', 'Noreste Ozono', 'Noreste dioxido de azufre',
       'Noreste dioxido de nitrogeno', 'Noreste monoxido de carbono',
       'Noreste PM10', 'Centro Ozono', 'Centro dioxido de azufre',
       'Centro dioxido de nitrogeno', 'Centro monoxido de carbono',
       'Centro PM10', 'Suroeste Ozono', 'Suroeste dioxido de azufre',
       'Suroeste dioxido de nitrogeno', 'Suroeste monoxido de carbono',
       'Suroeste PM10', 'Sureste Ozono', 'Sureste dioxido de azufre',
       'Sureste dioxido de nitrogeno', 'Sureste monoxido de carbono',
       'Sureste PM10'],
      dtype='object')

In [15]:
#Revisamos que todos los DataFrame tengan el mismo tamaño
for i in range(0,16):
    print(np.shape(DF_list2[i]))

(8760, 27)
(8760, 27)
(8760, 27)
(8784, 27)
(8760, 27)
(8760, 27)
(8760, 27)
(8784, 27)
(8781, 27)
(8760, 27)
(8760, 27)
(8784, 27)
(8760, 27)
(8760, 27)
(8760, 27)
(8784, 27)


Se observa que hay años en los que se tiene 8784 filas en lugar de 8760, esto es porque se trata de años bisiestos y tienen un día más, lo que equivale a 24 observaciones extra.

In [16]:
#Renombramos todas las columnas para no generar problemas al momento de concatenar
#DF_list2[0].columns

nombres_columnas=( "fecha"           ,        "hora"                ,          "noroeste_ozono"               ,
 "noroeste_dioxido_de_azufre"   , "noroeste_dioxido_de_nitrogeno", "noroeste_monoxido_de_carbono" ,
 "noroeste_pm10"                , "noreste_ozono"                , "noreste_dioxido_de_azufre"    ,
 "noreste_dioxido_de_nitrogeno" , "noreste_monoxido_de_carbono"  , "noreste_pm10"                 ,
 "centro_ozono"                 , "centro_dioxido_de_azufre"     , "centro_dioxido_de_nitrogeno"  ,
 "centro_monoxido_de_carbono"   , "centro_pm10"                  , "suroeste_ozono"               ,
 "suroeste_dioxido_de_azufre"   , "suroeste_dioxido_de_nitrogeno", "suroeste_monoxido_de_carbono" ,
 "suroeste_pm10"                , "sureste_ozono"                , "sureste_dioxido_de_azufre"    ,
 "sureste_dioxido_de_nitrogeno" , "sureste_monoxido_de_carbono"  , "sureste_pm10" )

for i in DF_list2:
    i.columns = nombres_columnas

In [17]:
#Concatenando todo en un mismo DataFrame
DF_all = pd.concat(DF_list2,axis=0)
DF_all=DF_all.reset_index(drop=True)
DF_all

fecha  hora  noroeste_ozono  noroeste_dioxido_de_azufre  \
0       01/01/2005   1.0            19.0                        19.0   
1       01/01/2005   2.0            19.0                        19.0   
2       01/01/2005   3.0            17.0                        19.0   
3       01/01/2005   4.0            16.0                        19.0   
4       01/01/2005   5.0            16.0                        19.0   
5       01/01/2005   6.0            16.0                        20.0   
6       01/01/2005   7.0            16.0                        21.0   
7       01/01/2005   8.0            14.0                        22.0   
8       01/01/2005   9.0            19.0                        22.0   
9       01/01/2005  10.0            46.0                        22.0   
10      01/01/2005  11.0            74.0                        23.0   
11      01/01/2005  12.0           103.0                        23.0   
12      01/01/2005  13.0           119.0                        23.0   
13      01/01/2005  14.0           122.0                        23.0   
14      01/01/2005  15.0           114.0                        23.0   
15      01/01/2005  16.0           102.0                        22.0   
16      01/01/2005  17.0            75.0                        22.0   
17      01/01/2005  18.0            53.0                        22.0   
18      01/01/2005  19.0            21.0                        22.0   
19      01/01/2005  20.0            25.0                        22.0   
20      01/01/2005  21.0            52.0                        22.0   
21      01/01/2005  22.0            57.0                        22.0   
22      01/01/2005  23.0            49.0                        22.0   
23      01/01/2005  24.0            36.0                        23.0   
24      02/01/2005   1.0            22.0                        24.0   
25      02/01/2005   2.0            34.0                        23.0   
26      02/01/2005   3.0            27.0                        22.0   
27      02/01/2005   4.0            26.0                        22.0   
28      02/01/2005   5.0            26.0                        21.0   
29      02/01/2005   6.0            20.0                        20.0   
...            ...   ...             ...                         ...   
140247  30/12/2020  19.0            24.0                         8.0   
140248  30/12/2020  20.0            26.0                         8.0   
140249  30/12/2020  21.0            28.0                         8.0   
140250  30/12/2020  22.0            27.0                         8.0   
140251  30/12/2020  23.0            27.0                         6.0   
140252  30/12/2020  24.0            26.0                         6.0   
140253  31/12/2020   1.0            26.0                         6.0   
140254  31/12/2020   2.0            26.0                         6.0   
140255  31/12/2020   3.0            26.0                         6.0   
140256  31/12/2020   4.0            25.0                         6.0   
140257  31/12/2020   5.0            26.0                         6.0   
140258  31/12/2020   6.0            30.0                         6.0   
140259  31/12/2020   7.0            28.0                         6.0   
140260  31/12/2020   8.0            21.0                         6.0   
140261  31/12/2020   9.0            20.0                         6.0   
140262  31/12/2020  10.0            26.0                         4.0   
140263  31/12/2020  11.0            29.0                         4.0   
140264  31/12/2020  12.0            31.0                         4.0   
140265  31/12/2020  13.0            34.0                         4.0   
140266  31/12/2020  14.0            40.0                         4.0   
140267  31/12/2020  15.0            41.0                         4.0   
140268  31/12/2020  16.0            35.0                         4.0   
140269  31/12/2020  17.0            31.0                         4.0   
140270  31/12/2020  18.0            28.0             

### Limpieza NaNs

In [18]:
#Revisando datos faltantes por columna
DF_all.isna().sum()

fecha                              21
hora                               21
noroeste_ozono                   1306
noroeste_dioxido_de_azufre         22
noroeste_dioxido_de_nitrogeno    1297
noroeste_monoxido_de_carbono     2027
noroeste_pm10                     174
noreste_ozono                    1299
noreste_dioxido_de_azufre          21
noreste_dioxido_de_nitrogeno     1331
noreste_monoxido_de_carbono      2073
noreste_pm10                      369
centro_ozono                     1319
centro_dioxido_de_azufre          101
centro_dioxido_de_nitrogeno      1366
centro_monoxido_de_carbono       2070
centro_pm10                       555
suroeste_ozono                   1236
suroeste_dioxido_de_azufre         42
suroeste_dioxido_de_nitrogeno    1261
suroeste_monoxido_de_carbono     2050
suroeste_pm10                     365
sureste_ozono                    1319
sureste_dioxido_de_azufre         126
sureste_dioxido_de_nitrogeno     1485
sureste_monoxido_de_carbono      2231
sureste_pm10

Se nota que hay columnas que presentan más datos faltantes que otras. Puesto que no podemos despreciar las mediciones de toda la fila solo porque una observación en una columna es nula, optamos en su lugar por llenar los valores vacíos con el valor promedio de la columna. De esta forma nos quedamos con suficientes datos para trabajar y mantenemos su valor promedio.

In [19]:
#Llenando los valores numéricos con el valor promedio
DF_all_subset=DF_all.loc[:,"noroeste_ozono":"sureste_pm10"]

for i in DF_all_subset.columns:
    DF_all[i]=DF_all[i].fillna(DF_all[i].mean())

In [20]:
#Corroborando que se hayan reemplazado todos los datos faltantes numéricos
DF_all.isna().sum()

fecha                            21
hora                             21
noroeste_ozono                    0
noroeste_dioxido_de_azufre        0
noroeste_dioxido_de_nitrogeno     0
noroeste_monoxido_de_carbono      0
noroeste_pm10                     0
noreste_ozono                     0
noreste_dioxido_de_azufre         0
noreste_dioxido_de_nitrogeno      0
noreste_monoxido_de_carbono       0
noreste_pm10                      0
centro_ozono                      0
centro_dioxido_de_azufre          0
centro_dioxido_de_nitrogeno       0
centro_monoxido_de_carbono        0
centro_pm10                       0
suroeste_ozono                    0
suroeste_dioxido_de_azufre        0
suroeste_dioxido_de_nitrogeno     0
suroeste_monoxido_de_carbono      0
suroeste_pm10                     0
sureste_ozono                     0
sureste_dioxido_de_azufre         0
sureste_dioxido_de_nitrogeno      0
sureste_monoxido_de_carbono       0
sureste_pm10                      0
dtype: int64

In [21]:
#Revisando valores nulos de fechas
DF_all[DF_all["fecha"].isna()]

fecha  hora  noroeste_ozono  noroeste_dioxido_de_azufre  \
78888   NaN   NaN       28.903253                    9.383715   
78889   NaN   NaN       28.903253                    9.383715   
78890   NaN   NaN       28.903253                    9.383715   
78891   NaN   NaN       28.903253                    9.383715   
78892   NaN   NaN       28.903253                    9.383715   
78893   NaN   NaN       28.903253                    9.383715   
78894   NaN   NaN       28.903253                    9.383715   
78895   NaN   NaN       28.903253                    9.383715   
78896   NaN   NaN       28.903253                    9.383715   
78897   NaN   NaN       28.903253                    9.383715   
78898   NaN   NaN       28.903253                    9.383715   
78899   NaN   NaN       28.903253                    9.383715   
78900   NaN   NaN       28.903253                    9.383715   
78901   NaN   NaN       28.903253                    9.383715   
78902   NaN   NaN       28.903253                    9.383715   
78903   NaN   NaN       28.903253                    9.383715   
78904   NaN   NaN       28.903253                    9.383715   
78905   NaN   NaN       28.903253                    9.383715   
78906   NaN   NaN       28.903253                    9.383715   
78907   NaN   NaN       28.903253                    9.383715   
78908   NaN   NaN       28.903253                    9.383715   

       noroeste_dioxido_de_nitrogeno  noroeste_monoxido_de_carbono  \
78888                      17.478565                     10.089902   
78889                      17.478565                     10.089902   
78890                      17.478565                     10.089902   
78891                      17.478565                     10.089902   
78892                      17.478565                     10.089902   
78893                      17.478565                     10.089902   
78894                      17.478565                     10.089902   
78895                      17.478565                     10.089902   
78896                      17.478565                     10.089902   
78897                      17.478565                     10.089902   
78898                      17.478565                     10.089902   
78899                      17.478565                     10.089902   
78900                      17.478565                     10.089902   
78901                      17.478565                     10.089902   
78902                      17.478565                     10.089902   
78903                      17.478565                     10.089902   
78904                      17.478565                     10.089902   
78905                      17.478565                     10.089902   
78906                      17.478565                     10.089902   
78907                      17.478565                     10.089902   
78908                      17.478565                     10.089902   

       noroeste_pm10  noreste_ozono  noreste_dioxido_de_azufre  \
78888      55.736244      31.219862                   7.527464   
78889      55.736244      31.219862                   7.527464   
78890      55.736244      31.219862                   7.527464   
78891      55.736244      31.219862                   7.527464   
78892      55.736244      31.219862                   7.527464   
78893      55.736244      31.219862                   7.527464   
78894      55.736244      31.219862                   7.527464   
78895      55.736244      31.219862                   7.527464   
78896      55.736244      31.219862                   7.527464   
78897      55.736244      31.219862                   7.527464   
78898      55.736244      31.219862                   7.527464   
78899      55.736244      31.219862                   7.527464   
78900      55.736244      31.219862                   7.527464   
78901      55.736244      31.219862                   7.527464   
78902      55.736244      31.219862         

In [22]:
#Revisando la última fecha correcta
DF_all.loc[78887]

fecha                            31/12/2013
hora                                     24
noroeste_ozono                            9
noroeste_dioxido_de_azufre               22
noroeste_dioxido_de_nitrogeno            15
noroeste_monoxido_de_carbono              7
noroeste_pm10                            48
noreste_ozono                             7
noreste_dioxido_de_azufre                11
noreste_dioxido_de_nitrogeno             15
noreste_monoxido_de_carbono              12
noreste_pm10                             48
centro_ozono                              1
centro_dioxido_de_azufre                 16
centro_dioxido_de_nitrogeno              16
centro_monoxido_de_carbono                9
centro_pm10                              48
suroeste_ozono                            5
suroeste_dioxido_de_azufre               15
suroeste_dioxido_de_nitrogeno            20
suroeste_monoxido_de_carbono             13
suroeste_pm10                            51
sureste_ozono                   

In [23]:
#Revisando fecha posterior a la última vacía
DF_all.loc[78909]

fecha                            01/01/2014
hora                                      1
noroeste_ozono                            8
noroeste_dioxido_de_azufre               18
noroeste_dioxido_de_nitrogeno            15
noroeste_monoxido_de_carbono              7
noroeste_pm10                            51
noreste_ozono                             8
noreste_dioxido_de_azufre                11
noreste_dioxido_de_nitrogeno             15
noreste_monoxido_de_carbono              13
noreste_pm10                             48
centro_ozono                              1
centro_dioxido_de_azufre                 15
centro_dioxido_de_nitrogeno              15
centro_monoxido_de_carbono                9
centro_pm10                              48
suroeste_ozono                            4
suroeste_dioxido_de_azufre               14
suroeste_dioxido_de_nitrogeno            18
suroeste_monoxido_de_carbono             14
suroeste_pm10                            53
sureste_ozono                   

In [24]:
#Se corrobora que el archivo de 2014 esté correcto
pd.read_csv("datos_IMECA/csv/imeca2014.csv")

Fecha  Hora  Noroeste Ozono  Noroeste dioxido de azufre  \
0     01/01/2014     1               8                          18   
1     01/01/2014     2               8                          18   
2     01/01/2014     3               5                          18   
3     01/01/2014     4               2                          18   
4     01/01/2014     5               3                          18   
5     01/01/2014     6               2                          18   
6     01/01/2014     7               2                          19   
7     01/01/2014     8               3                          20   
8     01/01/2014     9               5                          20   
9     01/01/2014    10               9                          20   
10    01/01/2014    11              23                          21   
11    01/01/2014    12              39                          21   
12    01/01/2014    13              57                          21   
13    01/01/2014    14              66                          21   
14    01/01/2014    15              65                          21   
15    01/01/2014    16              65                          21   
16    01/01/2014    17              63                          21   
17    01/01/2014    18              66                          21   
18    01/01/2014    19              45                          22   
19    01/01/2014    20              27                          25   
20    01/01/2014    21              25                          25   
21    01/01/2014    22              20                          24   
22    01/01/2014    23              17                          22   
23    01/01/2014    24              15                          21   
24    02/01/2014     1              14                          21   
25    02/01/2014     2              15                          20   
26    02/01/2014     3              19                          19   
27    02/01/2014     4              17                          19   
28    02/01/2014     5              15                          18   
29    02/01/2014     6               5                          18   
...          ...   ...             ...                         ...   
8730  30/12/2014    19              18                           5   
8731  30/12/2014    20              29                           5   
8732  30/12/2014    21              28                           5   
8733  30/12/2014    22              18                           4   
8734  30/12/2014    23              21                           4   
8735  30/12/2014    24              19                           5   
8736  31/12/2014     1              19                           5   
8737  31/12/2014     2              21                           7   
8738  31/12/2014     3              14                          11   
8739  31/12/2014     4              14                          15   
8740  31/12/2014     5               7                          15   
8741  31/12/2014     6               6                          16   
8742  31/12/2014     7               2                          17   
8743  31/12/2014     8               2                          18   
8744  31/12/2014     9               4                          18   
8745  31/12/2014    10               9                          18   
8746  31/12/2014    11              15                          19   
8747  31/12/2014    12              22                          19   
8748  31/12/2014    13              28                          19   
8749  31/12/2014    14              44                          22   
8750  31/12/2014    15              84                          22   
8751  31/12/2014    16             104                          22   
8752  31/12/2014    17             103                          22   
8753  31/12/2014    18              61                          22   
8754  31/12/2014    19              38                          24   
8755  31/12/2014    2

In [25]:
#Observando el archivo del 2013
pd.read_csv("datos_IMECA/csv/imeca2013.csv")

Fecha  Hora  Noroeste Ozono  Noroeste dioxido de azufre  \
0     01/01/2013   1.0             3.0                         2.0   
1     01/01/2013   2.0             3.0                         3.0   
2     01/01/2013   3.0             4.0                         3.0   
3     01/01/2013   4.0             5.0                         3.0   
4     01/01/2013   5.0             9.0                         4.0   
5     01/01/2013   6.0            13.0                         4.0   
6     01/01/2013   7.0             9.0                         4.0   
7     01/01/2013   8.0             7.0                         4.0   
8     01/01/2013   9.0             7.0                         5.0   
9     01/01/2013  10.0            20.0                         5.0   
10    01/01/2013  11.0            45.0                         5.0   
11    01/01/2013  12.0            55.0                         5.0   
12    01/01/2013  13.0            56.0                         5.0   
13    01/01/2013  14.0            63.0                         5.0   
14    01/01/2013  15.0            66.0                         5.0   
15    01/01/2013  16.0            56.0                         5.0   
16    01/01/2013  17.0            46.0                         5.0   
17    01/01/2013  18.0            36.0                         5.0   
18    01/01/2013  19.0            20.0                         5.0   
19    01/01/2013  20.0            12.0                         5.0   
20    01/01/2013  21.0             7.0                         5.0   
21    01/01/2013  22.0             6.0                         5.0   
22    01/01/2013  23.0            14.0                         5.0   
23    01/01/2013  24.0            14.0                         7.0   
24    02/01/2013   1.0            10.0                         7.0   
25    02/01/2013   2.0            15.0                         8.0   
26    02/01/2013   3.0            14.0                         9.0   
27    02/01/2013   4.0            11.0                        10.0   
28    02/01/2013   5.0             8.0                        10.0   
29    02/01/2013   6.0             6.0                        11.0   
...          ...   ...             ...                         ...   
8751  31/12/2013  16.0            76.0                        22.0   
8752  31/12/2013  17.0            75.0                        22.0   
8753  31/12/2013  18.0            48.0                        21.0   
8754  31/12/2013  19.0            22.0                        21.0   
8755  31/12/2013  20.0            15.0                        22.0   
8756  31/12/2013  21.0            10.0                        23.0   
8757  31/12/2013  22.0            15.0                        23.0   
8758  31/12/2013  23.0             3.0                        23.0   
8759  31/12/2013  24.0             9.0                        22.0   
8760         NaN   NaN             NaN                         NaN   
8761         NaN   NaN             NaN                         NaN   
8762         NaN   NaN             NaN                         NaN   
8763         NaN   NaN             NaN                         NaN   
8764         NaN   NaN             NaN                         NaN   
8765         NaN   NaN             NaN                         NaN   
8766         NaN   NaN             NaN                         NaN   
8767         NaN   NaN             NaN                         NaN   
8768         NaN   NaN             NaN                         NaN   
8769         NaN   NaN             NaN                         NaN   
8770         NaN   NaN             NaN                         NaN   
8771         NaN   NaN             NaN                         NaN   
8772         NaN   NaN             NaN                         NaN   
8773         NaN   NaN             NaN                         NaN   
8774         NaN   NaN             NaN                         NaN   
8775         NaN   NaN             NaN                         NaN   
8776         NaN   Na

Como se puede observar el archivo del 2013 contiene 8781 filas cuando debería contener 8760 ya que no es bisiesto. Es decir, tiene 21 filas vacías extra. Es necesario eliminar estas filas extra que no aportan información, esto se hará desde DF_all (el que acabamos de limpiar y contiene todos los arvhivos), puesto que son los únicos elementos vacíos que quedan.

In [26]:
#Tamaño antes de limpiar
DF_all.shape

(140277, 27)

In [27]:
#Eliminando filas con fecha vacía
DF_all=DF_all.dropna(axis="rows")

In [28]:
#Tamaño después de limpiar
DF_all.shape

(140256, 27)

In [29]:
#Corroborando que no queden elementos nulos
DF_all.isna().sum()

fecha                            0
hora                             0
noroeste_ozono                   0
noroeste_dioxido_de_azufre       0
noroeste_dioxido_de_nitrogeno    0
noroeste_monoxido_de_carbono     0
noroeste_pm10                    0
noreste_ozono                    0
noreste_dioxido_de_azufre        0
noreste_dioxido_de_nitrogeno     0
noreste_monoxido_de_carbono      0
noreste_pm10                     0
centro_ozono                     0
centro_dioxido_de_azufre         0
centro_dioxido_de_nitrogeno      0
centro_monoxido_de_carbono       0
centro_pm10                      0
suroeste_ozono                   0
suroeste_dioxido_de_azufre       0
suroeste_dioxido_de_nitrogeno    0
suroeste_monoxido_de_carbono     0
suroeste_pm10                    0
sureste_ozono                    0
sureste_dioxido_de_azufre        0
sureste_dioxido_de_nitrogeno     0
sureste_monoxido_de_carbono      0
sureste_pm10                     0
dtype: int64

### Transformación de Datos

In [30]:
#Revisando tipos de datos
DF_all.dtypes

fecha                             object
hora                             float64
noroeste_ozono                   float64
noroeste_dioxido_de_azufre       float64
noroeste_dioxido_de_nitrogeno    float64
noroeste_monoxido_de_carbono     float64
noroeste_pm10                    float64
noreste_ozono                    float64
noreste_dioxido_de_azufre        float64
noreste_dioxido_de_nitrogeno     float64
noreste_monoxido_de_carbono      float64
noreste_pm10                     float64
centro_ozono                     float64
centro_dioxido_de_azufre         float64
centro_dioxido_de_nitrogeno      float64
centro_monoxido_de_carbono       float64
centro_pm10                      float64
suroeste_ozono                   float64
suroeste_dioxido_de_azufre       float64
suroeste_dioxido_de_nitrogeno    float64
suroeste_monoxido_de_carbono     float64
suroeste_pm10                    float64
sureste_ozono                    float64
sureste_dioxido_de_azufre        float64
sureste_dioxido_

Observamos que hay que cambiar la fecha a formato `datetime`. La hora no nos va a ser tan útil en análisis posteriores puesto que queremos trabajar con promedios mensuales, sin embargo vamos a transformarla a formato `int` para que no genere confusiones. Todos los demás tipos de datos son correctos.

In [32]:
#Realizando el Casting correspondiente
casting={
    "fecha":"datetime64[ns]",
    "hora":"int"
}

DF_all=DF_all.astype(casting)

In [33]:
#Comprobando que se haya realizado correctamente el casting
DF_all.dtypes

fecha                            datetime64[ns]
hora                                      int32
noroeste_ozono                          float64
noroeste_dioxido_de_azufre              float64
noroeste_dioxido_de_nitrogeno           float64
noroeste_monoxido_de_carbono            float64
noroeste_pm10                           float64
noreste_ozono                           float64
noreste_dioxido_de_azufre               float64
noreste_dioxido_de_nitrogeno            float64
noreste_monoxido_de_carbono             float64
noreste_pm10                            float64
centro_ozono                            float64
centro_dioxido_de_azufre                float64
centro_dioxido_de_nitrogeno             float64
centro_monoxido_de_carbono              float64
centro_pm10                             float64
suroeste_ozono                          float64
suroeste_dioxido_de_azufre              float64
suroeste_dioxido_de_nitrogeno           float64
suroeste_monoxido_de_carbono            

Estos datos ya están limpios y transformados al tipo correcto, así que vamos a separarlos por tipo de contaminante y a guardarlos en archivos CSV para su análisis posterior.

In [35]:
#Separando la información por contaminante
DF_save = list()
palabra = ['ozono','dioxido_de_azufre','dioxido_de_nitrogeno','monoxido_de_carbono','pm10']
for i in range(0,5):
    DF_save.append(DF_all[['fecha','hora',f'noreste_{palabra[i]}',f'suroeste_{palabra[i]}',f'noroeste_{palabra[i]}',f'sureste_{palabra[i]}',f'centro_{palabra[i]}']])

In [36]:
DF_save[0]

fecha  hora  noreste_ozono  suroeste_ozono  noroeste_ozono  \
0      2005-01-01     1           10.0            26.0            19.0   
1      2005-01-01     2            7.0            19.0            19.0   
2      2005-01-01     3            5.0            23.0            17.0   
3      2005-01-01     4            6.0            21.0            16.0   
4      2005-01-01     5            8.0            23.0            16.0   
5      2005-01-01     6           13.0            25.0            16.0   
6      2005-01-01     7           11.0            35.0            16.0   
7      2005-01-01     8           19.0            31.0            14.0   
8      2005-01-01     9           10.0            22.0            19.0   
9      2005-01-01    10           20.0            41.0            46.0   
10     2005-01-01    11           73.0            80.0            74.0   
11     2005-01-01    12           94.0            98.0           103.0   
12     2005-01-01    13           86.0           124.0           119.0   
13     2005-01-01    14           83.0           161.0           122.0   
14     2005-01-01    15           88.0           172.0           114.0   
15     2005-01-01    16           73.0           122.0           102.0   
16     2005-01-01    17           74.0            97.0            75.0   
17     2005-01-01    18           62.0            57.0            53.0   
18     2005-01-01    19           60.0            45.0            21.0   
19     2005-01-01    20           60.0            35.0            25.0   
20     2005-01-01    21           31.0            40.0            52.0   
21     2005-01-01    22            6.0            48.0            57.0   
22     2005-01-01    23           17.0            35.0            49.0   
23     2005-01-01    24           13.0            35.0            36.0   
24     2005-02-01     1            4.0            34.0            22.0   
25     2005-02-01     2            4.0            45.0            34.0   
26     2005-02-01     3            4.0            44.0            27.0   
27     2005-02-01     4            5.0            32.0            26.0   
28     2005-02-01     5            4.0            31.0            26.0   
29     2005-02-01     6            4.0            35.0            20.0   
...           ...   ...            ...             ...             ...   
140247 2020-12-30    19           26.0            29.0            24.0   
140248 2020-12-30    20           25.0            29.0            26.0   
140249 2020-12-30    21           22.0            27.0            28.0   
140250 2020-12-30    22           26.0            26.0            27.0   
140251 2020-12-30    23           28.0            29.0            27.0   
140252 2020-12-30    24           26.0            30.0            26.0   
140253 2020-12-31     1           25.0            30.0            26.0   
140254 2020-12-31     2           26.0            29.0            26.0   
140255 2020-12-31     3           26.0            29.0            26.0   
140256 2020-12-31     4           26.0            29.0            25.0   
140257 2020-12-31     5           26.0            29.0            26.0   
140258 2020-12-31     6           24.0            29.0            30.0   
140259 2020-12-31     7           25.0            27.0            28.0   
140260 2020-12-31     8           24.0            21.0            21.0   
140261 2020-12-31     9           25.0            17.0            20.0   
140262 2020-12-31    10           27.0            24.0            26.0   
140263 2020-12-31    11           28.0            27.0            29.0   
140264 2020-12-31    12           31.0            30.0            31.0   
140265 2020-12-31    13           34.0            34.0            34.0   
140266 2020-12-31    14           36.0            37.0            40.0   
140267 2020-12-31    15           36.0            34.0            41.0   
140268 2020-12-31    16           36.0            33.0            35.0   
140269 2020-12-3

In [37]:
#Guardamos los datos limpios y transformados en archivos CSV

#Guardando archivos separados
nombres=['O3','SO2','NO2','CO','PM10']
for i in range(0,(len(nombres))):
    DF_save[i].to_csv(f'datos_IMECA/datos_limpios_imeca/{nombres[i]}.csv',index=False)
    
#Guardando también todo el DataFrame completo
DF_all.to_csv("datos_IMECA/datos_limpios_imeca/DF_all.csv",index=False)

A continuación vamos a calcular **promedios mensuales**, agrupando todos los datos por meses. Estos son los datos que serán más útiles para nuestro análisis.

In [38]:
#Eliminando hora ya que agruparemos por meses
DF_all = DF_all.drop(['hora'], axis=1)

In [39]:
#Agrupando por mes
#Se indexa basado en tiempo, lo que permite un mejor manejo con datos temporales 
DF_all_mensual=DF_all.resample('M', on='fecha').mean()

In [40]:
DF_all_mensual

noroeste_ozono  noroeste_dioxido_de_azufre  \
fecha                                                    
2005-01-31       35.766129                   19.295699   
2005-02-28       34.148810                   12.596726   
2005-03-31       34.479839                   12.700269   
2005-04-30       38.872222                   22.769444   
2005-05-31       41.360215                   11.712366   
2005-06-30       30.511111                    9.711111   
2005-07-31       30.771505                   13.916667   
2005-08-31       30.123656                    9.798387   
2005-09-30       32.184722                   13.331944   
2005-10-31       29.918011                   15.138441   
2005-11-30       31.150000                   18.561111   
2005-12-31       32.153226                   19.061828   
2006-01-31       31.241935                   14.337366   
2006-02-28       30.968750                   12.880952   
2006-03-31       31.381720                   10.317204   
2006-04-30       37.773611                    8.483333   
2006-05-31       35.556452                   11.834677   
2006-06-30       32.875000                    8.451389   
2006-07-31       26.506720                    9.182796   
2006-08-31       30.720430                   12.075269   
2006-09-30       30.438889                    8.787500   
2006-10-31       27.760753                    9.380376   
2006-11-30       31.312500                   13.204167   
2006-12-31       28.059140                   13.131720   
2007-01-31       26.479839                   11.254032   
2007-02-28       33.836310                   10.901786   
2007-03-31       35.231183                    9.752688   
2007-04-30       32.465278                    8.512500   
2007-05-31       33.118280                    8.301075   
2007-06-30       32.606944                   10.187500   
...                    ...                         ...   
2018-07-31       28.426553                    6.729839   
2018-08-31       26.020509                    6.142473   
2018-09-30       25.305646                    6.145833   
2018-10-31       24.239725                    5.534946   
2018-11-30       28.791354                    6.250000   
2018-12-31       27.215835                   10.333333   
2019-01-31       24.925252                    9.377688   
2019-02-28       23.502593                    7.904762   
2019-03-31       32.043228                    9.478495   
2019-04-30       40.933424                    6.797222   
2019-05-31       38.887054                   11.255376   
2019-06-30       29.928854                   11.148611   
2019-07-31       27.238815                    8.822581   
2019-08-31       29.173865                   11.884409   
2019-09-30       26.542026                   11.861111   
2019-10-31       24.918098                   12.608871   
2019-11-30       24.731049                   18.856944   
2019-12-31       22.690428                   19.606183   
2020-01-31       21.851588                   19.135753   
2020-02-29       27.826660                   19.814655   
2020-03-31       28.083551                   12.524194   
2020-04-30       33.572313                   10.274144   
2020-05-31       36.115679                   10.524194   
2020-06-30       27.324956                   19.584722   
2020-07-31       26.584115                    9.456989   
2020-08-31       24.614595                   10.618280   
2020-09-30       25.583692                   14.509722   
2020-10-31       24.339317                   14.639785   
2020-11-30       22.882975                   10.983333   
2020-12-31       25.839057                   13.819892   

            noroeste_dioxido_de_nitrogeno  noroeste_monoxido_de_carbono  \
fecha                                                                     
2005-01-31                      22.278226                     18.849462   
2005-02-28                      20.175595                     16.983631   
2005-03-31                      19.801075                

Se observa que la agrupación se realizó correctamente, la fecha representativa es el último día del mes. A continuación guardamos este dataframe completo y separado por contaminantes.

In [41]:
#Separando la información por contaminante
DF_save = list()
palabra = ['ozono','dioxido_de_azufre','dioxido_de_nitrogeno','monoxido_de_carbono','pm10']
for i in range(0,5):
    DF_save.append(DF_all_mensual[[f'noreste_{palabra[i]}',f'suroeste_{palabra[i]}',f'noroeste_{palabra[i]}',f'sureste_{palabra[i]}',f'centro_{palabra[i]}']])

In [42]:
DF_save[0]

noreste_ozono  suroeste_ozono  noroeste_ozono  sureste_ozono  \
fecha                                                                      
2005-01-31      34.497312       42.629032       35.766129      40.125000   
2005-02-28      34.537202       42.840774       34.148810      39.645833   
2005-03-31      37.178763       41.713710       34.479839      40.633065   
2005-04-30      43.033333       46.236111       38.872222      43.913889   
2005-05-31      45.501344       50.325269       41.360215      47.463710   
2005-06-30      30.472222       33.940278       30.511111      31.745833   
2005-07-31      30.895161       37.323925       30.771505      34.518817   
2005-08-31      28.014785       38.044355       30.123656      33.883065   
2005-09-30      31.018056       41.372222       32.184722      37.162500   
2005-10-31      29.102151       41.077957       29.918011      35.668011   
2005-11-30      32.330556       42.441667       31.150000      34.530556   
2005-12-31      34.587366       43.420699       32.153226      36.946237   
2006-01-31      30.848118       42.185484       31.241935      33.025538   
2006-02-28      32.910714       45.102679       30.968750      35.180060   
2006-03-31      32.107527       41.564516       31.381720      32.458333   
2006-04-30      36.290278       48.137500       37.773611      38.677778   
2006-05-31      35.443548       44.149194       35.556452      37.655914   
2006-06-30      31.038889       40.836111       32.875000      34.298611   
2006-07-31      25.309140       34.346774       26.506720      28.689516   
2006-08-31      29.456989       39.646505       30.720430      32.064516   
2006-09-30      27.973611       38.041667       30.438889      31.461111   
2006-10-31      27.927419       36.588710       27.760753      29.481183   
2006-11-30      29.745833       40.077778       31.312500      30.736111   
2006-12-31      27.801075       35.909946       28.059140      28.513441   
2007-01-31      27.270457       36.473118       26.479839      28.360215   
2007-02-28      34.699405       44.620536       33.836310      34.133929   
2007-03-31      33.049731       46.600806       35.231183      35.653226   
2007-04-30      34.556944       46.161111       32.465278      34.462500   
2007-05-31      35.876344       47.192204       33.118280      35.840054   
2007-06-30      32.369444       44.866667       32.606944      35.584722   
...                   ...             ...             ...            ...   
2018-07-31      29.179755       34.816730       28.426553      28.213018   
2018-08-31      25.966341       30.693837       26.020509      26.077265   
2018-09-30      24.644996       29.472708       25.305646      24.464927   
2018-10-31      23.914674       27.033128       24.239725      21.555760   
2018-11-30      26.648690       33.893128       28.791354      25.511124   
2018-12-31      27.265452       33.750251       27.215835      27.812075   
2019-01-31      26.107441       31.552852       24.925252      27.292149   
2019-02-28      25.612884       28.345799       23.502593      25.457047   
2019-03-31      33.454539       37.023901       32.043228      32.865134   
2019-04-30      42.365830       46.693541       40.933424      41.146871   
2019-05-31      42.564001       47.033491       38.887054      45.765907   
2019-06-30      31.551468       35.101461       29.928854      32.430568   
2019-07-31      27.921071       31.636585       27.238815      28.291796   
2019-08-31      29.669002       34.909472       29.173865      30.809792   
2019-09-30      27.658274       32.296131       26.542026      28.355083   
2019-10-31      25.743813       28.132190       24.918098      24.730843   
2019-11-30      27.030191       30.521732       24.731049      29.086785   
2019-12-31      24.957361       30.784435       22.690428      27.076743   
2020-01-31      23.596098       27.215305       21.851588      24.584925   
2020-02-29      28.195025       32.766734       27.826660   

In [43]:
#Guardamos los datos limpios y transformados en archivos CSV

#Guardando archivos separados
nombres=['O3','SO2','NO2','CO','PM10']
for i in range(0,(len(nombres))):
    DF_save[i].to_csv(f'datos_IMECA/promedios_mensuales/{nombres[i]}_mensual.csv',index=True)
    
#Guardando también todo el DataFrame completo
DF_all_mensual.to_csv("datos_IMECA/promedios_mensuales/DF_all_mensual.csv",index=True)

### Resultados preliminares

A continuación trabajaremos con nuestros datos limpios y transformados para intentar responder algunas preguntas de investigación, sin un análisis tan profundo ni visualización, ya que esto se abordará en el siguiente módulo.

-   ¿En qué meses se tiene la máxima concentración de cada contaminante?

    

    
-   ¿Cuál es la zona que presenta mayor contaminación, en general?

    
-   ¿Cuál es el contaminante con más concentración por año (del 2005- 2020)?

    
-   ¿Cuál fue y cuándo se registró la concentración más baja durante 2020?

    
-   ¿Cuál fue y cuándo se registró la mayor concentración durante el periodo de confinamiento?

    
-   ¿El periodo de confinamiento coincide con una mejor calidad de aire?

    
-   ¿Cuál fue el contaminante con menores cambios en el período 2005-2020?

**¿Cuál es el promedio histórico (considerando los años 2005-2020) de concentración de cada contaminante por zona?**

In [44]:
#Promedio histórico
DF_all.set_index("fecha").mean(axis=0).round(2)

noroeste_ozono                   28.90
noroeste_dioxido_de_azufre        9.38
noroeste_dioxido_de_nitrogeno    17.48
noroeste_monoxido_de_carbono     10.09
noroeste_pm10                    55.74
noreste_ozono                    31.22
noreste_dioxido_de_azufre         7.53
noreste_dioxido_de_nitrogeno     16.09
noreste_monoxido_de_carbono       9.96
noreste_pm10                     69.31
centro_ozono                     25.97
centro_dioxido_de_azufre          6.44
centro_dioxido_de_nitrogeno      18.36
centro_monoxido_de_carbono       10.90
centro_pm10                      52.78
suroeste_ozono                   36.57
suroeste_dioxido_de_azufre        5.01
suroeste_dioxido_de_nitrogeno    15.20
suroeste_monoxido_de_carbono      8.51
suroeste_pm10                    40.04
sureste_ozono                    33.09
sureste_dioxido_de_azufre         4.68
sureste_dioxido_de_nitrogeno     15.97
sureste_monoxido_de_carbono      10.08
sureste_pm10                     51.64
dtype: float64

Los promedios históricos se observan bastante bien considerando que debajo de las 50 unidades IMECA la calidad del aire es buena.

**¿Hay algún mes de algún año en el que se haya obtenido la calificación de calidad “Extremadamente mala”?**

Recordemos que la calidad _Extremadamente mala_ corresponde a los valores mayores a 200

In [46]:
#Calidad extremadamente mala
extremadamente_mala=[]

for column in DF_all_mensual.columns:
    extremadamente_mala.append(DF_all_mensual[DF_all_mensual[column]>200])
    
extremadamente_mala=pd.concat(extremadamente_mala,axis=0)
extremadamente_mala

Empty DataFrame
Columns: [noroeste_ozono, noroeste_dioxido_de_azufre, noroeste_dioxido_de_nitrogeno, noroeste_monoxido_de_carbono, noroeste_pm10, noreste_ozono, noreste_dioxido_de_azufre, noreste_dioxido_de_nitrogeno, noreste_monoxido_de_carbono, noreste_pm10, centro_ozono, centro_dioxido_de_azufre, centro_dioxido_de_nitrogeno, centro_monoxido_de_carbono, centro_pm10, suroeste_ozono, suroeste_dioxido_de_azufre, suroeste_dioxido_de_nitrogeno, suroeste_monoxido_de_carbono, suroeste_pm10, sureste_ozono, sureste_dioxido_de_azufre, sureste_dioxido_de_nitrogeno, sureste_monoxido_de_carbono, sureste_pm10]
Index: []

[0 rows x 25 columns]

Se observa que no hay ningún registro con calidad _Extremadamente mala_ de acuerdo a los valores mensuales reportados.

#### ¿Cuál es el contaminante con más concentración por año (del 2005- 2020)?

In [105]:
#Agrupamos por año
DF_all_anual=DF_all.resample('Y', on='fecha').mean()
#Obtenemos el índice del valor máximo por fila
DF_all_anual.idxmax(axis = 1)

fecha
2005-12-31    noreste_pm10
2006-12-31    noreste_pm10
2007-12-31    noreste_pm10
2008-12-31    noreste_pm10
2009-12-31    noreste_pm10
2010-12-31    noreste_pm10
2011-12-31    noreste_pm10
2012-12-31    noreste_pm10
2013-12-31    noreste_pm10
2014-12-31    noreste_pm10
2015-12-31    noreste_pm10
2016-12-31    noreste_pm10
2017-12-31    noreste_pm10
2018-12-31    noreste_pm10
2019-12-31    noreste_pm10
2020-12-31    noreste_pm10
Freq: A-DEC, dtype: object

Se puede observar que el contaminante con más concentración por año fueron las partículas en suspensión (PM10) en la misma zona de noroeste.